src/mlops/orchestration/orchestration_caller.py

```diff
@@ -0,0 +1,363 @@
+    return None, None
+
+
+def _test_passed_orchestrator(orc_endpoint: str, orc: str) -> bool:
+    """
+    test the user passed orchestrator and endpoint information
+    to ensure it is correct and will intgrate with the pipeline
+    orchestration wrapper
+
+    :param orc_endpoint: the endpoint for producers/consumers to connect to
+    :type orc_endpoint: str
+    :param orc: the name of the orchestrator
+    :type orc: str
+
+    :return: True if valid information passed false otherwise
+    :rtype: bool
+    """
+    # validate orchestrator string
+    if orc is None:
+        logging.info("Orchestrator name not passed")
+        return False
+    if orc.lower() not in APPROVED_ORCHESTRATORS:
+        logging.info(f"{orc} is not an approved orchestrator for the pipeline")
+        return False
+
+    # validate orchestrator endpoint
+    if orc_endpoint is None:
+        logging.info("Orchestrator endpoint not passed")
+        return False
+    # try to ping orchestrator to make sure it is up
+    command = ["ping", "-c", "1", orc.lower()]
+    if subprocess.call(command) == 0:
+        return True
+    command = ["ping", "-c", "1", orc_endpoint]
+    if subprocess.call(command) == 0:
+        return True
+    logging.info(f"Unable to ping orchestrator at {orc_endpoint}")
+    return False
+
+
+# ------------------------------------------------
+# Message Classes
+# ------------------------------------------------
+
+
+class Producer:
+    def __init__(
+        self,
+        topic: str or List[str] = None,
+        orchestrator: str = None,
+        orchestrator_endpoint: str = None,
+        **kwargs,
+    ):
+        """
+        Initialize the higher level Producer class exposed to the user
+
+        :param topic: the event bucket this producer writes to. Should not
+                contain any white spaces. These will be removed.
+        :type topic: str or List[str]
+        :param orchestrator: the name of the orchestrator in use
+        :type orchestrator: str
+        :param orchestrator_endpoint: The endpoint of the orchestrator
+        :type orchestrator_endpoint: str
+        :param `**kwargs`: any additional configuration requirements desired
+                for the Producer. Keys based on underlying orchestration
+                requirements. Look at the docs for the orchestrator wrapper
+                to determine desired requirements
+        :type `**kwargs`: Any
+        """
+        # determine underlying orchestrator
+        if _test_passed_orchestrator(orchestrator_endpoint, orchestrator):
+            self.orchestrator = orchestrator.lower()
+            self.orchestrator_endpoint = orchestrator_endpoint
+        else:
+            (
+                self.orchestrator,
+                self.orchestrator_endpoint,
+            ) = _get_orchestrator()
+
+        # connect to orchestrator and create the producer
+        if self.orchestrator == "kafka":

```

Comment:
This goes back to my "interface based design" comment from earlier in the review. You can hide a lot of this complexity with dispatching to the appropriate function at call-time by adopting the following strategy:

- Define the interface for the underlying orchestration backend, codify this some type e.g. `OrchestrationProducer`
- Have each of the backends you want _implement_ this interface. E.g:

```python
class KafkaProducer(OrchestrationProducer):
  ...
```

- You can enforce the interface with Python metaclasses, which makes it much harder to make mistakes when programming a new backend.
- Create an instance of the backend with the constructor for the `Producer`, e.g. `self.underlying = `KafkaProducer()`
- Now all of the methods on `Producer` can simply delegate to the underlying implementation, without any conditional logic:

```python
def send(...):
  self.underlying.send(...)
```